examples/cluster/plot_kmeans_assumptions.py

```diff
@@ -4,66 +4,129 @@
 
 """
 
 # Author: Phil Roth <mr.phil.roth@gmail.com>
+#         Arturo Amor <david-arturo.amor-quiroz@inria.fr>
 # License: BSD 3 clause
 
-import numpy as np
-import matplotlib.pyplot as plt
+# %%
+# Data generation
+# ---------------
+#
+# The function :func:`~sklearn.datasets.make_blobs` generates isotropic gaussian
+# blobs. To obtain anisotropic (elliptical) gaussian blobs one has to define a
+# linear `transformation`.
 
-from sklearn.cluster import KMeans
+import numpy as np
 from sklearn.datasets import make_blobs
 
-plt.figure(figsize=(12, 12))
-
 n_samples = 1500
 random_state = 170
+transformation = [[0.60834549, -0.63667341], [-0.40887718, 0.85253229]]
+
 X, y = make_blobs(n_samples=n_samples, random_state=random_state)
+X_aniso = np.dot(X, transformation)  # Anisotropic blobs
+X_varied, y_varied = make_blobs(
+    n_samples=n_samples, cluster_std=[1.0, 2.5, 0.5], random_state=random_state
+)  # Unequal variance
+X_filtered = np.vstack(
+    (X[y == 0][:500], X[y == 1][:100], X[y == 2][:10])
+)  # Unevenly sized blobs
 
-# Incorrect number of clusters
-y_pred = KMeans(n_clusters=2, n_init="auto", random_state=random_state).fit_predict(X)
+# %%
+# Fit models and plot results
+# ---------------------------
 
-plt.subplot(221)
-plt.scatter(X[:, 0], X[:, 1], c=y_pred)
-plt.title("Incorrect Number of Blobs")
+import matplotlib.pyplot as plt
+from sklearn.cluster import KMeans
 
-# Anisotropicly distributed data
-transformation = [[0.60834549, -0.63667341], [-0.40887718, 0.85253229]]
-X_aniso = np.dot(X, transformation)
-y_pred = KMeans(n_clusters=3, n_init="auto", random_state=random_state).fit_predict(
-    X_aniso
-)
+common_params = {
+    "n_init": "auto",
+    "random_state": random_state,
+}
 
-plt.subplot(222)
-plt.scatter(X_aniso[:, 0], X_aniso[:, 1], c=y_pred)
-plt.title("Anisotropicly Distributed Blobs")
+fig, axs = plt.subplots(nrows=2, ncols=2, figsize=(12, 12))
 
-# Different variance
-X_varied, y_varied = make_blobs(
-    n_samples=n_samples, cluster_std=[1.0, 2.5, 0.5], random_state=random_state
-)
-y_pred = KMeans(n_clusters=3, n_init="auto", random_state=random_state).fit_predict(
-    X_varied
-)
+y_pred = KMeans(n_clusters=2, **common_params).fit_predict(X)
+axs[0, 0].scatter(X[:, 0], X[:, 1], c=y_pred)
+axs[0, 0].set_title("Incorrect Number of Blobs")
+
+y_pred = KMeans(n_clusters=3, **common_params).fit_predict(X_aniso)
+axs[0, 1].scatter(X_aniso[:, 0], X_aniso[:, 1], c=y_pred)
+axs[0, 1].set_title("Anisotropically Distributed Blobs")
 
-plt.subplot(223)
-plt.scatter(X_varied[:, 0], X_varied[:, 1], c=y_pred)
-plt.title("Unequal Variance")
+y_pred = KMeans(n_clusters=3, **common_params).fit_predict(X_varied)
+axs[1, 0].scatter(X_varied[:, 0], X_varied[:, 1], c=y_pred)
+axs[1, 0].set_title("Unequal Variance")
+
+y_pred = KMeans(n_clusters=3, **common_params).fit_predict(X_filtered)
+axs[1, 1].scatter(X_filtered[:, 0], X_filtered[:, 1], c=y_pred)
+axs[1, 1].set_title("Unevenly Sized Blobs")
+
+plt.show()
```

Comment:
a small discussion to explain that when we don't do a break the assumptions, we don't get the expected result.